src/bin/cargo/commands/remove.rs

```diff
@@ -154,28 +254,72 @@ fn gc_workspace(workspace: &Workspace<'_>, dependencies: &[String]) -> CargoResu
     Ok(())
 }
 
-/// Get whether or not a dependency is depended upon in a workspace.
-fn dep_in_workspace(dep: &str, members: &[LocalManifest]) -> bool {
-    members.iter().any(|manifest| {
-        manifest.get_sections().iter().any(|(_, table)| {
-            table
-                .as_table_like()
-                .unwrap()
-                .get(dep)
-                .and_then(|t| t.get("workspace"))
-                .and_then(|v| v.as_bool())
-                .unwrap_or(false)
-        })
-    })
+/// Check whether or not a package ID spec matches any dependencies.
+fn spec_has_match(
+    spec: &PackageIdSpec,
+    dependencies: &[Dependency],
+    config: &Config,
+) -> CargoResult<bool> {
+    for dep in dependencies {
+        match dep.source_id(config)? {
+            MaybeWorkspace::Other(source_id) => {
+                if spec.name().as_str() != &dep.name {
+                    continue;
+                }
+
+                let version_matches = match (spec.version(), dep.version()) {
+                    (Some(v), Some(vq)) => semver::VersionReq::parse(vq)?.matches(v),
+                    (Some(_), None) => false,
+                    (None, _) => true,
```

Comment:
nit: I'd make the `_` explicitly a `None`

Reply:
To have `(None, None)` and `(None, Some)`? Both of those are handled the same (no spec version means all other versions, existent or not, are handled), which is why they're grouped together.

Reply:
I tend to prefer that to be explicit, even if its `(None, Some(_)) | (None, None)`.  This makes it easier to see what cases to reason about and helps when refactoring